### ISSUE TITLE
Allow running Background once per feature

### ISSUE THREAD

**Issue by jacek99** [cucumber/cucumber#594]
Right now, as per typical Gherkin standards, a Background section is run once for every Scenario.

Although understandable, this really breaks down when running complex integration tests with expensive setup. 

If we know all the Scenarios in a Feature file are just read-only (e.g. testing GET commands of a REST API), then tearing down and setting it all up again between every scenario adds *significant* execution time.

This is especially evident on Scenario Outlines / Examples (e.g. when sending different variations of an invalid JSON payload to a REST API to ensure input validation tests). Having to setup/teardown between every execution Scenario Outline is painfully inefficient.

Please consider *some* way to allow running a Background only *once* per feature.

Maybe new heading "Feature Background:" (vs just "Background")...that would be easiest.

This would make a world of difference for shops like ours that have complex integration test suites for REST APIs in microservices. Really, a *WORLD* of difference.

This is the biggest single complaint I hear about Cucumber from every single developer on my team.

Thank you in advance.

**Comment by aslakhellesoy** [comment]
You can make sure a step definition's body is run only once like this:

```java
import io.cucumber.java.api.en.Given;

public class MyStepdefs {
    private static boolean somethingExpensiveDone = false;

    @Given("something expensive")
    public void somethingExpensive() throws InterruptedException {
        if(!somethingExpensiveDone) {
            somethingExpensiveDone = true;
            doSomethingExpensive();
        }
    }

    private void doSomethingExpensive() throws InterruptedException {
        Thread.sleep(3000);
    }
}
```

Would that work for you @jacek99?

**Comment by jacek99** [comment]
@aslakhellesoy unfortunately, that would be totally insufficient. We have complex setups that involve 10-15 entities (and multiple REST calls) to set up base data for some test condition (common requirement in a multi-tenant application where most configuration & data is different from one customer to another).

I've seen one very complex BDD where the Background section is 300 lines.

And then multiple Scenarios that test various queries against that setup.

So a **single** "something expensive" step is really not a good solution here, unfortunately.

"Feature Background"" or "Setup:" is really what we need....

**Comment by aslakhellesoy** [comment]
Could you please explain what's insufficient with the workaround I posted?

> I've seen one very complex BDD where the Background section is 300 lines.

This is not what Cucumber is designed for. It's not a testing tool, it's a communication tool, and 300 lines is too long for communication. See https://cucumber.ghost.io/blog/the-worlds-most-misunderstood-collaboration-tool/

**Comment by jacek99** [comment]
@aslakhellesoy the setup varies from one Feature file to another, we'd have to write a Given "something expensive" step separately fro every such complex Feature and it would hide what sort of data was set up for the Scenario.

Cucumber is a testing too. I am sorry...I totally disagree with your comment. I have used it for testing AND communication for over 10 years now, across 3 different language implementations.

**Comment by aslakhellesoy** [comment]
You can call it whatever you want, but it doesn't change the fact that I designed Cucumber to be a tool for simple, readable executable specifications with slow, heavy setup, used in a "test-after" context (as opposed to "test-first".

We won't be adding this feature, but I have created a new ticket to collect various extensions people are requesting: https://github.com/cucumber/cucumber/issues/595

When/if such an extension API is implemented, you should be able to implement a custom extension yourself.

Sorry I can't be of more help.

**Comment by jacek99** [comment]
I regret you closing this. We're a heavy Cucumber user and it has been great for us. We are pushing the envelope on using Cucumber for REST microservice integration tests...for complex, multi-tenant applications with complex data conditions.

I've had the **exact** same requirement in all of my last 4 jobs when using Cucumber and we struggled with it every time.

**Comment by aslakhellesoy** [comment]
It conflicts with the overall philosophy of Cucumber. As the lead of this project I need to be conservative about what goes into the tool. 

I believe that adding this is encouraging people to use Cucumber differently from what I believe is good practice. That's my prerogative.

**Comment by jacek99** [comment]
I respect that.

But also ask you to consider understanding that it may be very usable in other scenarios, very different than its initial design.

I'd go with Cucumber over writing integration tests in Java/JUnit any time of the week. Massive improvement in every aspect.

**Comment by aslakhellesoy** [comment]
Yes, I understand that. The thing is, there is over a million people using Cucumber, and only half a dozen people maintaining it on a regular basis (in their spare time, unpaid).

If you want this feature, you'll have to implement it yourself. Because it's something I don't think should be part of the official release, I want to provide an API where you can plug it in.

I can't promise when/if this extension API will be available, but I'd be happy to guide anyone who volunteers to build it.

**Comment by enkessler** [comment]
If I might add my thoughts here...

From a purely semantic perspective, a `Background` clearly applies to every use case in a feature. Execution-wise, however, both interpreting it to only happen once per feature or interpreting it to happen once per use case are perfectly reasonable approaches. As it happened, Cucumber went with the 'once per use case' implementation. So @jacek99 wanting a way to use the alternative interpretation is not unreasonable.

But @aslakhellesoy is also perfectly justified in his concerns about the _motivation_ for this alternate implementation. Gherkin is a communication tool that facilitates easy to understand _specifications_. Cucumber is a (for lack of a better term) 'testing tool' that allows those specifications to be _executed_. We all love executable specifications, right?

Conflict like I am seeing in this thread seems to come from the fact that a big portion of the user base (okay, let's be honest, the majority) focusing on the _executable_ half of 'executable specification' while the designers often focus on the _specification_ half. **But we can be both**. `Background`s that can either be 'once per feature' or 'once per use case' is a legitimate goal*. The question is how to get there. It _can_ be done today (as demonstrated by @aslakhellesoy 's example code) but are there other ways? Ways that might facilitate both 'specification' and 'execution'. Let's resolve that question, if we can.


**TLDR**: Feel free to continue debating on the best way to achieve an alternate implementation of `Background` (new keyword, flags in step defs or hooks, etc.) but, please, actually have that debate.



*I'm going to assume that the feature is otherwise a good feature. Yes, users can mis-use Gherkin and have a 300 step `Background` but they will do that whether or not that background runs once or many times. Let's not throw the baby out with the bathwater.

**Comment by enkessler** [comment]
Additionally: Because a `Background` is just as meaningful to the reader whether it is executed once per feature or once per use case, instead of adding a new keyword or having to do step def/hook magic, perhaps we could simply have a command line flag that controls which in which mode backgrounds run?

`--repeat-background` and `--no-repeat-background`?

**Comment by jacek99** [comment]
@enkessler not a good option, since this varies by Feature. We have simple feature files where the current Background functionality is perfect, as the subsequent Scenarios mutate the state...so tearing down/up is required.

I would say only 10-15% of our BDDs have such a heavy set up time that we need this "run background once per feature"...but those BDDs account for most of the test suite execution time ....
